### DISCUSSION TITLE
Running MacOS test on SPM to generate Kotlin...  missing "test-results" folder

### DISCUSSION THREAD

**Discussion by davidakoontz** [skiptools/skip#285] [Q&A]
I'm attempting to port DeckKit to Skip it is a SPM module.  I've modified a dozen line and rewritten a few files to hopefully work in Kotlin.

When setting up to target MacOS and run the test - which in turn runs gradle - getting this missing folder.

failed - The expected test output folder did not exist, which may indicate that the gradle process encountered a build error or other issue. Missing folder: /Users/david/Library/Developer/Xcode/DerivedData/deckkit-skip-enzcmhnfbyicljfgevwhpmvzekcr/SourcePackages/checkouts/skip/Skip/build/deckkit-skip.output/DeckKitTests/skipstone/DeckKit/.build/DeckKit/test-results

Sure enought that **test-results** folder does not exist.

<img width="1369" alt="Screenshot 2024-11-09 at 1 28 51 PM" src="https://github.com/user-attachments/assets/e4df928f-3e9a-40ac-953e-addd747c6d73">

Does this ring any bells about what may be wrong?  Thanks,
David

**Comment by marcprux** [comment]
That error usually indicates that something else went wrong earlier in the build or test process. If you open the debug area of Xcode and look in the console section, can you look and see if there is an error message?

**Reply by davidakoontz** [reply]
I will attempt it again... switch back to that project...  Here is the test run... all errors...
<img width="1143" alt="Screenshot 2024-11-10 at 12 59 13 PM" src="https://github.com/user-attachments/assets/cba07686-97bc-453a-a511-028a66f30b81">
the build log has just a few yellow warnings... rest is all green checks...
<img width="1210" alt="Screenshot 2024-11-10 at 1 02 58 PM" src="https://github.com/user-attachments/assets/d9814def-4029-4298-b096-cb659d93b07d">

I will try your suggestion in a bit...

**Comment by davidakoontz** [comment]
I'm not to confident in my build env for a SPM module and getting Android Studio & Device up and running...

With an App - there is an Android folder...  But with just a Module there is NO Android folder - so what project do I open with Android Studio?
What is REQUIRED for building the Module?

**Reply by davidakoontz** [reply]
I'm attempting to follow the Docs...

https://skip.tools/docs/gettingstarted/#framework_development

**Comment by davidakoontz** [comment]
OK - reran the DeckKit-Skip UnitTest to fire up the Kotlin build process for the SPM ...  here is the Console Log... lots of errors...

One of them hates that DeckKit has camel-case... and got Java-fied into deck.kit...  I've been confused by this quirk since day one with Skip...  any advice?

Namespace 'deck.kit' is used in multiple modules and/or libraries: :DeckKit, :SkipUnit, :SkipFoundation, :SkipLib. Please ensure that all modules and libraries have a unique namespace. For more information, See https://developer.android.com/studio/build/configure-app-module#set-namespace

[ConsoleLog-TestRun-DeckKit-Skip.txt](https://github.com/user-attachments/files/17692564/ConsoleLog-TestRun-DeckKit-Skip.txt)

Here is my DeckKit-Skip project tree...

[Tree-or-DeckKit-Skip.txt](https://github.com/user-attachments/files/17692580/Tree-or-DeckKit-Skip.txt)

**Comment by davidakoontz** [comment]
After reading some of https://developer.android.com/build/configure-app-module#set-namespace 
About the Android Name Space...  I went looking for the build.gradle.kts - didn't find one until I inspected the hidden .build folder...

Way down deep in the folder is a generated build.gradle.kts with a namespace value of "group as String" - that seems WRONG to me!

android {
    namespace = group as String

<img width="1175" alt="Screenshot 2024-11-10 at 1 55 43 PM" src="https://github.com/user-attachments/assets/238dcab6-8419-4946-9179-cc7200077ebf">

**Comment by davidakoontz** [comment]
Here is the Console log from another #3 Attempt:  started back at step 1...  skip init...

Built when no new files....  then copied my DeckKit SPM source into place... ran test (Android build)... lots of errors in Console log.
[ConsoleLog-DeckKit-attempt3.txt](https://github.com/user-attachments/files/17693053/ConsoleLog-DeckKit-attempt3.txt)

**Reply by marcprux** [reply]
Errors like these make it sound like you aren't depending on SkipUI, or that your files aren't importing SwiftUI:

```
e: file:///Users/david/Library/Developer/Xcode/DerivedData/deck-kit-dnhgzicxdhbuoraiefsswgthgapf/SourcePackages/plugins/deck-kit.output/DeckKitTests/skipstone/DeckKit/src/main/kotlin/deck/kit/Card.kt:17:17 Unresolved reference 'compose'.
e: file:///Users/david/Library/Developer/Xcode/DerivedData/deck-kit-dnhgzicxdhbuoraiefsswgthgapf/SourcePackages/plugins/deck-kit.output/DeckKitTests/skipstone/DeckKit/src/main/kotlin/deck/kit/Card.kt:20:13 Unresolved reference 'ui'.
e: file:///Users/david/Library/Developer/Xcode/DerivedData/deck-kit-dnhgzicxdhbuoraiefsswgthgapf/SourcePackages/plugins/deck-kit.output/DeckKitTests/skipstone/DeckKit/src/main/kotlin/deck/kit/Card.kt:22:13 Unresolved reference 'model'.
e: file:///Users/david/Library/Developer/Xcode/DerivedData/deck-kit-dnhgzicxdhbuoraiefsswgthgapf/SourcePackages/plugins/deck-kit.output/DeckKitTests/skipstone/DeckKit/src/main/kotlin/deck/kit/Card.kt:29:26 Unresolved reference 'View'.
```

Try adding the dependency to your `Package.swift` and let us know if the errors persist.

**Reply by davidakoontz** [reply]
Do I have to modify the Package.swift file for SkipUI?
Many of the files explicitly import SwiftUI;  some do not just import Foundations;  what would I need to do in that case?

Here's Package.swift...  I not quite following the implications.  I'm not clear on what I need to add...

<img width="1368" alt="Screenshot 2024-11-10 at 9 03 51 PM" src="https://github.com/user-attachments/assets/1615d9de-5555-401b-87b8-830d5ba3b966">

**Reply by davidakoontz** [reply]
I added (correctly I hope) skip-ui to dependencies - didn't seem to help.

    dependencies: [
        .package(url: "https://source.skip.tools/skip.git", from: "1.1.18"),
        .package(url: "https://source.skip.tools/skip-foundation.git", from: "1.0.0"),
        .package(url: "https://source.skip.tools/skip-ui.git", from: "1.0.0")
    ],

**Reply by davidakoontz** [reply]
After running the Test... looking in the Swift/Xcode IDE at Swift file... I see Errors that look like it's never seen a SWIFT file.

<img width="1384" alt="Screenshot 2024-11-10 at 9 57 44 PM" src="https://github.com/user-attachments/assets/786eefdb-cdbe-4e69-9ce5-bda9e8d558dc">

**Reply by marcprux** [reply]
> I added (correctly I hope) skip-ui to dependencies - didn't seem to help.

Does your target also add the SkipUI dependency?  E.g.:

```swift
.target(name: "Deckkit", dependencies: [
    .product(name: "SkipUI", package: "skip-ui")
], 
plugins: [.plugin(name: "skipstone", package: "skip")]),
```

You can model this off of https://github.com/skiptools/skip-kit/blob/main/Package.swift, or one of the many other sample repositories we have at https://github.com/skiptools/.

**Reply by davidakoontz** [reply]
Oh - there's more to adding the dependencies; an experience I've not had... I must also expand the targets dependencies...  KNOWLEDGE is a wonderful thing.

Thanks Marc.  I'm feeling like I'm learning now.  Added the target dependencies on skip-ui  (like the example Package you sent - thanks for that, examples are priceless! ) ...
And with that one small change my build went from 425 errors to a staggering 176 errors!  Real progress...  I bet if I fix up the Test target & dependencies I may get close to Zero errors!

I truly appreciate the expert guidance!

**Reply by davidakoontz** [reply]
Nope!  I was wrong... (again)!

Those 176 errors are not resolved by adding the skip-ui to the Test target dependencies.

**Reply by davidakoontz** [reply]
if you happen to be following this saga...  I've gotten the error count down to about 20!  

Getting closer....  now much of the issues are around Identifiable Array values...  which I think Swift is auto generating an 'id' property... but Java is not ... so the item.id  <<== error Unresolved ref 'id'

Do I need to make that id property explicit for Skip to make the Java code transpire correctly?
